### PR TITLE
[Coop] Convert System.Security.Cryptography.RNGCryptoServiceProvider.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

**Comment by jaykrell** [comment]
Looks like relevant CI failure. Will address later.

**Comment by jaykrell** [comment]
@monojenkins build failes

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/utils/mono-rand.c [hunk 1]
```diff
@@ -93,19 +93,19 @@ mono_getrandom (guchar *buffer, gint buffer_size, gint flags, MonoError *error)
...
 static gint
-mono_getentropy (guchar *buffer, gint buffer_size, MonoError *error)
+mono_getentropy (guchar *buffer, intptr_t buffer_size, MonoError *error)
 {
 	static gint getentropy_works = 1;
 	gint count = 0;
 	gint len;
 	gint err;
 
+	error_init (error);
```
Comment (luhenry) [inline mono/utils/mono-rand.c]: Remove this `error_init`. `mono_getentropy` is only called by `mono_rand_try_get_bytes` which already calls `error_init`.

File: mono/utils/mono-rand.c [hunk 2]
```diff
@@ -53,18 +53,18 @@ static gint file = -1;
...
    getrandom() is retried if it failed with EINTR: interrupted by a signal. */
 static gint
-mono_getrandom (guchar *buffer, gint buffer_size, gint flags, MonoError *error)
+mono_getrandom (guchar *buffer, intptr_t buffer_size, gint flags, MonoError *error)
 {
 	static gint getrandom_works = 1;
 	gint count = 0;
 	gint err;
 
+	error_init (error);
```
Comment (luhenry) [inline mono/utils/mono-rand.c]: Remove this `error_init`. `mono_getrandom` is only called by `mono_rand_try_get_bytes` which already calls `error_init`.

File: mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs [hunk 3]
```diff
@@ -94,31 +98,33 @@ private void Check ()
 		private static extern bool RngOpen ();
 		
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
-		private static extern IntPtr RngInitialize (byte[] seed);
+		unsafe private static extern IntPtr RngInitialize (byte* seed, IntPtr seed_length);
```
Comment (luhenry) [inline mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs]: Pass the `seed_length` as a `gint32`.

File: mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs [hunk 4]
```diff
@@ -94,31 +98,33 @@ private void Check ()
...
 
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
-		private static extern IntPtr RngGetBytes (IntPtr handle, byte[] data);
+		unsafe private static extern IntPtr RngGetBytes (IntPtr handle, byte* data, IntPtr data_length);
```
Comment (luhenry) [inline mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs]: Pass the `data_length` as a `gint32`.

Comment (jaykrell) [reply mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs]: `uintptr_t mono_array_handle_length (MonoArrayHandle arr);`
but I used intptr_r as compromise to preserve previous signedness in existing code.

**Comment by jaykrell** [comment]
Later changes maybe crashing, will see.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs (hunk 4 above)
Comment (jaykrell) [reply mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs]: We also have:
```
#ifdef MONO_BIG_ARRAYS
typedef uint64_t mono_array_size_t;
#else
typedef uint32_t mono_array_size_t;
#endif
```
however the ifdef does not translate into the BCL interface -- we only build it once without ifdef. So you have to pick one type that works either way, with either configuration of C and with C#, so the wider type is what works. The smaller type would be silently truncated and silent truncation of integers is dangerous and not worth much risk.

**Comment by jaykrell** [comment]
@monojenkins build pkg with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

File: configure.ac [hunk 5]
```diff
@@ -40,7 +40,7 @@ MONO_VERSION_BUILD=`echo $VERSION | cut -d . -f 3`
 # This can be reset to 0 when Mono's version number is bumped
 # since it's part of the corlib version (the prefix '1' in the full
 # version number is to ensure the number isn't treated as octal in C)
-MONO_CORLIB_COUNTER=9
+MONO_CORLIB_COUNTER=0
```
Comment (jaykrell) [inline configure.ac]: Should be 10 but wasn't working, trying 1.0.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Subtle break on Windows where an invalid parameter is getting a different exception than expected. I don't see where it would have originated in the original. It is obvious where it originates in the revision. Will probably have to build/run on windows and break on prepare_exception.

File: mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs (hunk 4 above)
Comment (luhenry) [reply mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs]: https://msdn.microsoft.com/en-us/library/system.array.length%28v=vs.110%29.aspx. `Array.Length` returns `int` which is equivalent to `gint32`.

Comment (jaykrell) [reply mcs/class/corlib/System.Security.Cryptography/RNGCryptoServiceProvider.cs]: But what about Mono's large array feature? Worthless? Only for embedding uses that aren't interacting with C#/IL?

**Comment by jaykrell** [comment]
Still an error to figure out in a deliberately degenerate test case.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
@monojenkins build Linux x64

**Comment by jaykrell** [comment]
review? Also checking with @baulig, I think he is kinda but not quite obsoleting this code, and conflicting on approximately one line. Maybe he goes first.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
CI passing. Approve?

**Comment by jaykrell** [comment]
@monojenkins build failed

**Pull request merged**